splicemachine/spliceengine
# DB-11905 Additional mechanism to throttle the streaming (new PR after revert)

## Short Description
The additional channel between StgreamListener and StreamableRDD to manage throttling 

## Long Description
Original PR https://github.com/splicemachine/spliceengine/pull/5519
**Reason for revert and fix**: locally the execution of ITs failed with Too many open files: java.io.IOException. The first fix has corrected this problem, but some ITs were hanging. 
The fix:
1. The channel resources are closed in OlapStreamListener now.
2. The reason for hanging was the usage of sync close method, which can wait forever if a channel is closed already on another side. 

What is changed, compare to the original approved PR:

1. close method added:

```
public void close() throws IOException {
        try {
            if (channelFuture != null)
                channelFuture.channel().closeFuture();
            if (workerGroup != null)
                workerGroup.shutdownGracefully();
        } catch (Exception e) {
            LOG.warn("Could not close channel resources", e);
        }
    }
```
    
2. close method is called from StreamableRDD on the end of submit:
```
} finally {
            streamingExecutor.shutdown();
            olapStreamListener.close();
}
```

The throttling mechanism we have currently, see StreamableRDD.submit(), is to run 2 concurrent jobs at a time on the spark side, each sending rows from 2 partitions to the region server. If the buffer for a particular partition is full, it would cause back-pressure, and prevent Spark from sending more rows for that partition. However, if the number of rows in the partition is smaller than the buffer size, then there is another mechanism to hold Spark from sending all partitions to the region server. 
There is additional channel from StreamableRDD to StreamListener. It uses the same hostname/port, which is passed with submit command and used to connect from ResultStreamer.
When the StreamListener has more than 3 * DEFAULT_PARALLEL_PARTITIONS in the queue, it sends the message PauseStream to StreamableRDD to pause to submit new partitions to Spark. If the consumption starts again, StreamListener sends the message ContinueStream to resume the partions submitting.

The functionality can be disabled by configuration parameter **spark.result.streaming.throttle.enabled**, by default is true.

## How to test
```
create table a ( i int);
insert into a values 1,2,3;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
```

```
get cursor c as 'select * from a --splice-properties useSpark=true
order by i';
next c;

ps -ef | grep 'com.splicemachine.test.SpliceTestPlatform' | less

jmap -histo:live [PID] | grep 'PartitionState'

```
It should not be bigger than 12 on the standalone cluster

Tu run ITs locally for regression:
mvn -B install -Pcore,ee,dbaas3.0 '-Dtest=!*' -DfailIfNoTests=false
